yt_dlp/extractor/niconico.py

```diff
@@ -351,7 +351,26 @@ def ping():
 
         return info_dict, heartbeat_info_dict
 
-    def _extract_format_for_quality(self, video_id, audio_quality, video_quality, dmc_protocol):
+    def _get_dms_manifest_url(self, info_dict):
+        formats = info_dict['url'].split(':')[1].split('/')[1:2]
+        payload = json.dumps({
+            'outputs': [formats]
+        }).encode("utf-8")
+        api_data = self._download_json(
+            'https://nvapi.nicovideo.jp/v1/watch/%s/access-rights/hls?actionTrackId=%s' % (info_dict['id'], traverse_obj(info_dict, ('_api_data', 'client', 'watchTrackId'))), info_dict['id'],
+            note='Requesting Manifest Url', errnote='Unable to fetch data',
+            data=payload,
+            headers={
+                'Accept-Encoding': 'br',
+                'Content-Type': 'application/json',
+                'X-Request-With': 'https://www.nicovideo.jp',
+                'X-Access-Right-Key': traverse_obj(info_dict, ('_api_data', 'media', 'domand', 'accessRightKey')),
+                'X-Frontend-Id': '6',
+                'X-Frontend-Version': '0',
+            }).get('data')
+        return api_data.get('contentUrl')
```

Comment:
You probably don't need it, but just in case, here it is
DMS must create a session specifying video and audio quality to download
This method should not be used for all formats, as a new session is created for each request.
The reason this method is implemented in extractor instead of downloader is that the heartbeat generation method was implemented in extractor.